### DISCUSSION TITLE
Water level sensor using esp8266

### DISCUSSION THREAD

**Discussion by ramkithepower** [arendst/Tasmota#15620] [Q&A] [answered]
Hi All, I have esp8266 nodemcu board with which i am trying to put in a water level sensor. The one that i am using is the one below.
![61pgsHmACsL _SY879_](https://user-images.githubusercontent.com/30769165/168428076-c32150c4-5118-4c50-9aca-71621917c51f.jpg)

The idea is to drop them into the tank tied to a pipe, such that 5v line is at the bottom and then other 4 pins  tied at various levels of water. When the water starts filling the contact is closed at defined levels which should then set the switches which I have configured in D0,D1,D2,D3 to ON appropriately and then that information across to openhab on MQTT. 5v for the bottom most pin is taken from the board itself.
In the 'Configure Template' menu, i have configured the respective GPIO pins has Switch and then numbered them from 1 to 4. To test this i am directly shorting the pins from 5v, but I can get only "SWITCH4" configured at D0 pin to work. All others are not working.

If you see the console log below, the Switch4 alone is set to OFF in the start, all others are set to ON. Also when i short the pins with 5v, Switch 4 alone works correctly, others are not responding. Any thoughts?

```
12:37:54.378 MQT: tele/tasmota_699C7E/STATE = {"Time":"2022-05-14T12:37:54","Uptime":"0T00:20:10","UptimeSec":1210,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"sachin","BSSId":"10:50:72:40:BC:3E","Channel":1,"Mode":"11n","RSSI":58,"Signal":-71,"LinkCount":1,"Downtime":"0T00:00:04"}}
12:37:54.384 MQT: tele/tasmota_699C7E/SENSOR = {"Time":"2022-05-14T12:37:54","Switch1":"ON","Switch2":"ON","Switch3":"ON","Switch4":"OFF"}
```
<img width="1260" alt="Screenshot 2022-05-14 at 5 12 23 PM" src="https://user-images.githubusercontent.com/30769165/168424226-9af9d30f-392f-4b43-a8c3-02639bf20a56.png">
<img width="398" alt="Screenshot 2022-05-14 at 5 13 06 PM" src="https://user-images.githubusercontent.com/30769165/168424228-b84ec005-4646-4e07-8465-ec57d5ae8fd9.png">

**Comment by barbudor** [comment, marked as answer]
> The one that i am using is the one below
I see no reference to a water leak sensor so it is difficult to understand how it is working

> directly shorting the pins from 5v
Are you talking about connecting an ESP GPIO to +5V : this is about killing the ESP. The ESP is a 3.3V chip and GPIOs are NOT 5V tolerant.

Beware that some GPIOs of an ESP8266 have special functions and may be already connected to pull-up/pull-down on the board 
https://tasmota.github.io/docs/Expanding-Tasmota/#gpio-overview
- GPIO0 and GPIO2 have pull-ups and GPIO15 have pull-down. Those levels MUST be respected at boot otherwise the ESP will not boot Tasmota
- GPIO1 and GPIO3 are used for the serial link and connected to the USP/Serial chip on the NodeMCU. You may use them as far as USB is not connected and have disabled serial console with `SerialLog 0`

**Reply by ramkithepower** [reply]
I have attached the sensor's pic above now. Can you please think about it?
Also the sensor that is attached to GPIO0 seems to be working fine. The others are not.

**Reply by sfromis** [reply]
GPIO0 should have an external pullup on ESP8266 boards. By default, Tasmota will activate a weak internal pullup on other pins, but that could be too weak. In that case, try with external pullup on your other GPIO pins. As barbudor mentioned, you should also have an external pullup on GPIO2, does it work if you use that pin? I'd avoid GPIO16 unless having a real need, as this pin should have a built-in pulldown, which may not work well with all devices.

**Reply by barbudor** [reply]
So the so called "sensor" is no more than an fully passive inoxydable pin ?
Or is there any electronics between those probes and the ESP ?
May be pointing to a page of the product or documentation ?

But as pointed by sfromis GPIO0 works fine, it is probably because you need to add external pull ups. The pull up on GPIO0 and GPIO2 is generally 10-12K

**Reply by ramkithepower** [reply]
After @sfromis @barbudor  pointed out the pulldown factor on GPIO16, the only switch that worked( I was initially wrong, GPIO0 is not working), i tried connecting one of the pin to GPIO15, which also was pulled to ground and that switch works too now. Does it mean i have to add pull down resistors for other pins as well. As per this link [](https://randomnerdtutorials.com/esp8266-pinout-reference-gpios/) I think I can use D1, D2, D5, D6 pins for my purpose with pull down added???

**Reply by sfromis** [reply]
Yes, that may mean that pulldown resistors are needed, but as first step I'd try no longer having the automatic pullup of Tasmota enabled, by configuring pins as `Switch_n` instead of just `Switch`. Anyway, you should also config as `Switch_n` when adding pulldown resistors.
